src/argilla/client/feedback/dataset/remote/base.py

```diff
@@ -90,154 +95,24 @@ def __parse_record(self, record: "FeedbackItemModel") -> RemoteFeedbackRecord:
-                raise IndexError(
-                    f"Index {start if start < 0 else stop} is out of range, dataset has {num_records} records."
-                )
-            limit = stop - start
-            if limit < 0:
-                raise ValueError("Negative slice bounds are not supported.")
-            elif limit < FETCHING_BATCH_SIZE:
-                offsets = [start]
-                limits = [limit]
-            else:
-                offsets = list(range(start, stop, FETCHING_BATCH_SIZE))
-                limits = [FETCHING_BATCH_SIZE] * len(offsets)
-                if stop % FETCHING_BATCH_SIZE != 0:
-                    offsets[-1] = stop - (stop % FETCHING_BATCH_SIZE) + 1
-                    limits[-1] = (stop % FETCHING_BATCH_SIZE) - 1
-        elif isinstance(key, int):
-            if key < 0:
-                key += num_records
-            if key < 0 or key >= num_records:
-                raise IndexError(f"Index {key} is out of range, dataset has {num_records} records.")
-            offsets = [key]
-            limits = [1]
-        else:
-            raise TypeError("Only `int` and `slice` are supported as index.")
-
-        records = []
-        for offset, limit in zip(offsets, limits):
-            fetched_records = datasets_api_v1.get_records(
-                client=self._client,
-                id=self._dataset_id,
-                offset=offset,
-                limit=limit,
-            ).parsed
-            records.extend([self.__parse_record(record) for record in fetched_records.items])
-        return records[0] if isinstance(key, int) else records
-
-    @allowed_for_roles(roles=[UserRole.owner, UserRole.admin])
-    def __iter__(self) -> Iterator[RemoteFeedbackRecord]:
-        """Iterates over the `FeedbackRecord`s of the current `FeedbackDataset` in Argilla."""
-        current_batch = 0
-        while True:
-            batch = datasets_api_v1.get_records(
-                client=self._client,
-                id=self._dataset_id,
-                offset=FETCHING_BATCH_SIZE * current_batch,
-                limit=FETCHING_BATCH_SIZE,
-            ).parsed
-            for record in batch.items:
-                yield self.__parse_record(record)
-            current_batch += 1
-
-            if len(batch.items) < FETCHING_BATCH_SIZE:
-                break
+    @abstractmethod
+    def _fetch_records(self, offset: int, limit: int) -> "FeedbackRecordsModel":
+        pass
 
-    @allowed_for_roles(roles=[UserRole.owner, UserRole.admin])
     def add(
         self,
-        records: Union[FeedbackRecord, Dict[str, Any], List[Union[FeedbackRecord, Dict[str, Any]]]],
+        records: Union["FeedbackRecord", Dict[str, Any], List[Union["FeedbackRecord", Dict[str, Any]]]],
         show_progress: bool = True,
     ) -> None:
-        """Pushes a list of `FeedbackRecord`s to Argilla.
+        raise NotImplementedError(f"`records.add` not implemented for `{self.__class__.__name__}`.")
```

Comment:
could we define the `add`, `update`, and `delete` as abstract methods and let each implementation notify with a more friendly message? For example, for the filtered datasets, I can imagine a message informing users that filtered datasets do not support adding or deleting data.
```suggestion
    @abstract
    def add(...)
        pass
```